File: 103__Binary_Tree_Zigzag_Level_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        queue = [root]
        result = []
        reverse = False
        while queue:
            next = []
            temp = []
            for u in queue:
                if u.left:
                    next.append(u.left)
                if u.right:
                    next.append(u.right)
                temp.append(u.val)
            if not reverse:
                result.append(temp)
            else:
                result.append(temp[::-1])
            queue = next
            reverse = not reverse
        return result

File: Java/103__Binary_Tree_Zigzag_Level_Order_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    List<List<Integer>> result = new ArrayList<>();

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        helper(root, 0);
        return result;
    }

    private void helper(TreeNode root, int level) {
        if (root == null) return;
        if (result.size() <= level)
            result.add(new ArrayList<>());
        if (level % 2 == 0)
            result.get(level).add(root.val);
        else
            result.get(level).add(0, root.val);
        helper(root.left, level + 1);
        helper(root.right, level + 1);
    }
}
